[3 - 4단계 방탈출 예약 대기] 멍구(이유영) 미션 제출합니다.

YuyoungRhee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
    [https://prolog.techcourse.co.kr/studylogs/4329](https://prolog.techcourse.co.kr/studylogs/4329)
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

3-4단계 요구사항과 더불어 필요한 부분은 일부 리팩터링을 진행했습니다.

- Service에 존재하는 정책적인 규칙 분리(ReservationTimePolicy, ReservationPolicy)
    - [refactor(ReservationService): 예약 가능 조건 정책 분리](https://github.com/woowacourse/spring-roomescape-waiting/commit/8252dcf1c768c58e90a2b62130af4f3842ac0d8e)
    - [refactor(ReservationTimeService): 시간에 관한 정책을 별도의 클래스로 분리](https://github.com/woowacourse/spring-roomescape-waiting/commit/71ae3928659c6615f430c1ace4d9271b4331b8df)
- Controller 및 빠진 테스트 보강

### waiting은 Reservation의 상태? 새로운 도메인?

1-2단계에서는 간단히 '예약'이라는 상태를 보여주기만 하면 됐으므로 Reservation의 status 상태를 통해 함께 관리했습니다. 그러나 대기에 관련된 비즈니스 로직들이 추가되면서"예약"과 "대기" 를 분리할 필요가 있는지 고민하게 되었습니다.

**현재 대기와 예약이 별개의 행위인가? : 🔺**
- controller 엔드포인트는 다르나 service로직이 동일하거나 유사합니다. 
- Reservation 엔티티 하나로 대기와 예약을 모두 처리하고 있습니다.
    - 생성: status만 다르게 설정
    - 삭제: 동일

**대기만의 고유한 속성이 존재하는가?: ❌**
- 대기와 예약이 동일한 속성들(member, date, time, theme)을 공유하고 있습니다.
- 대기 번호는 조회 시 계산되는 정보이며, DB 및 도메인 속성이 아닙니다.

**대기와 예약은 다른 책임으로 볼 수 있는가?: 🔺**
- 현재 설계에서는 다른 책임이 아닙니다. 대기와 예약 모두 Reservation의 책임입니다. 
- 그러나 해당 도메인을 설계하는 입장으로서는 본질적으로 다른 개념이라고 느껴졌습니다. 즉 얼마든지 설계를 다르게 할 수 있다고 판단되는 두 가지 개념이라고 생각했습니다.

**결론**: 현재 코드에서는 분리해야될만큼의 분리된 개념은 아니며, Reservation에서 충분히 처리 가능하다. 그러나 본질적으로는 다른 개념으로 느껴지며 얼마든지 분리될 수 있는 개념이라 생각된다.

Reservation의 책임이 커지는 것 같아 분리를 고민하고 있었습니다만, 지노의 의견도 들어보고 싶어 우선 기존 설계를 지키며 구현해보았습니다! 

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

YuyoungRhee committed:
refactor: 예약이 존재하는 시간대 조회 로직을 데이터베이스단에서 처리하도록 변경

YuyoungRhee committed:
refactor(ReservationTimeService): 시간에 관한 정책을 별도의 클래스로 분리

YuyoungRhee committed:
fix(RegisterMemberRequest): 파라미터 순서 불일치 문제 수정

YuyoungRhee committed:
refactor: get vs find 네이밍 컨벤션 반영

YuyoungRhee committed:
refactor(ReservationService): 예약 가능 조건 정책 분리

YuyoungRhee committed:
refactor: 예약 초기 데이터 날짜에 맞게 넣어주도록 변경

YuyoungRhee committed:
refactor: 1+N 발생 Fetch join으로 해결

YuyoungRhee committed:
refactor: 불필요한 TODO 및 주석 삭제

YuyoungRhee committed:
fix: 클라이언트 코드를 3단계에 맞게 수정

YuyoungRhee committed:
refactor(controller): view Controller 패키지 구분

YuyoungRhee committed:
feat: 예약 대기 요청 API 구현

YuyoungRhee committed:
feat: 예약 대기 번호 프론트에 표시

YuyoungRhee committed:
refactor: 예약 대기 번호를 서비스에서 계산하는 것으로 변경

YuyoungRhee committed:
refactor: 대기 예약 순번을 JPA 쿼리에서 직접 계산하도록 변경

- 서비스 로직의 복잡도와 DB 접근 횟수를 줄임

YuyoungRhee committed:
refactor: delteById 로직을 adminReservationController로 옮김

YuyoungRhee committed:
feat: 4단계 예약 관리 프론트 코드 추가

YuyoungRhee committed:
refactor(ReservationController): find -> get 네이밍 변경

YuyoungRhee committed:
feat: 예약 대기 목록 조회 api 구현

YuyoungRhee committed:
feat: 예약 대기 삭제 api 구현

YuyoungRhee committed:
refactor: get - find 네이밍 컨벤션 모두 반영

YuyoungRhee committed:
refactor(exception): NotFoundException 하나로 통일

YuyoungRhee committed:
refactor(service): 트랜잭션이 필요한 로직에 @Transactional 추가

YuyoungRhee committed:
refactor(service): 예약 가능 여부 메서드를 반대 결과로 리팩터링하여 가독성 향상

YuyoungRhee committed:
feat(ReservationCreationService): 대기 가능 검증 추가

YuyoungRhee committed:
fix(ReservationService): 대기 취소에서 자동 예약 발생하는 오류 수정

YuyoungRhee committed:
test: 테스트 보강

YuyoungRhee committed:
refactor(MemberController): 회원가입 성공 응답을 201로 변경

YuyoungRhee committed:
fix(RegisterMemberParam): 인자 순서 오류 수정

YuyoungRhee committed:
feat(exception): 잡히지 않는 예외는 500처리하도록 추가

YuyoungRhee committed:
fix: 예약 목록에서 대기까지 같이 보이는 현상 수정

- 예약 목록은 RESERVED 상태의 예약만 보이도록 수정

YuyoungRhee committed:
refactor: 메서드 네이밍 오타 수정

YuyoungRhee committed:
refactor: 가독성을 위한 메서드 순서 변경

YuyoungRhee committed:
feat: 불필요한 예외 로그 출력 방지를 위한 NoResourceFoundException 핸들러 추가

yuyoungrhee commented on src/main/java/roomescape/config/SampleReservationInitializer.java:
```diff
@@ -0,0 +1,86 @@
+
+@Component
+@Profile("local")
+public class SampleReservationInitializer implements CommandLineRunner {
```
날짜에 따라 인기 테마 데이터가 바뀝니다. 예시 데이터 날짜를 항상 바꿔주는 번거로움을 줄이고자 "예약" 데이터에 한해 자바 코드로 저장하도록 했습니다.

yuyoungrhee commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -13,15 +24,15 @@ public class Reservation {
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY)
     @JoinColumn(name = "member_id", nullable = false)
     private Member member;
 
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY)
     @JoinColumn(name = "time_id", nullable = false)
     private ReservationTime time;
 
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY)
     @JoinColumn(name = "theme_id", nullable = false)
     private Theme theme;
```
연관관계에 어노테이션에 대해 더 공부해보면서 fetch 전략에 대해 알게 되었습니다. 
`@ManyToOne`의 기본 전략은 Eager이며, Reservation을  조회할 떄 연관된 객체들이 "즉시 로딩"되는 문제가 있습니다. 즉시 로딩의 문제점은, 필요가 없음에도 무거운 객체들이 무조건 함께 조회된다는 점입니다. 이를 개발자가 통제하고 예상할 수 있도록 프로그래밍하기 위하여 Lazy로 설정해주었습니다.

hyunssooo replied:
지연 로등에 대해서 학습하셨군요! 지연 로딩에 대해서는 단점이 없었나요? 어떤 기준으로 Eager를 사용하고 LAZY를 사용하나요?

yuyoungrhee replied:
지연로딩의 장점은 객체를 필요한 순간에 실제 db에서 가져올 수 있다는 전략에 있습니다. 다만 연관 객체 접근 시 추가 쿼리가 발생한다는 점은 단점으로 작용할 수 있고, 그로 인해 성능 저하나 N+1 문제가 발생할 수 있습니다. (다만 단일객체 조회, 페이징된 컬렉션 조회 시 성능 문제가 있지는 않다고 판단됩니다)

그러나 LAZY / EAGER에서 각각 **최선의 설계를 했다고 가정했을 때** LAZY는 EAGER에 비해서 단점이 없다고 생각합니다.
LAZY에서 신경써야 하는 부분이 더 많은 것은 맞습니다. 항상 연관되어있는 객체가 필요한 시점에 로딩된다는 것을 인지하고, N+1 문제 등을 미리 신경쓴다면 LAZY가 더 유연하며 성능 제어(불필요한 로딩 방지)에도 도움된다고 생각합니다.

각각 N+1 문제를 해결한다면 쿼리 횟수 차이는 없는 것 같습니다.
**ManyToOne**
- LAZY: fetch join 시 쿼리 1회
- EAGER: 기본적으로 join 쿼리 1회 / 명시적으로 fetch join 활용도 가능

**OneToOne**
- LAZY: fetch join 시 1회
- EAGER: 1회

**OneToMany**
- LAZY: FetchMode.SUBSELECT or BatchSize or dto projection - 쿼리 2회 / row 중복 없음
- EAGER: FetchMode.SUBSELECT - 쿼리 2회 / row 중복 없음

---
따라서 저는 연관관계에서 모두 LAZY를 사용하며, N+1 문제를 인지하고 명시적으로 해결하는 방법으로 기준을 세우려 합니다. 

예외: 정말 모든 기능에서 항상 연관 객체가 필요하고, 매번 즉시 로딩이 자연스러운 도메인이 있다면 그때는 EAGER도 고려할 수 있다고 생각합니다.
- (이런 경우의 예시가 생각나는 건 아직 없습니다)

yuyoungrhee commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,94 +1,102 @@
 
 @Service
+@Transactional(readOnly = true)
 public class ReservationService {
```
현재 서비스에서 조회 결과로 모두 ReservationResult를 포함하고 있습니다.
여기서 화면에 필요 없는 정보(member의 email, theme의 thumbnail, description 등)을 모두 포함합니다.

이를 해결하는 방법에는 2가지가 있을 것 같습니다.

1. **Controller에서** Result -> Response로 변환할 때 해당 api가 요청하는 필요한 정보만 가공한다
2. **Service에서** Entity -> Result로 변환할 때 해당 유즈케이스가 필요로하는 정보만 가공한다.

Service는 요청이 원하는 정보에 대해서 너무 세부적으로 알 필요는 없는 것인지, 반대로 Service가 요청에 대해 정확하게 그 정보를 가공할 책임이 있는 것인지 고민입니다. 현재 제 생각은, 어떤 정보를 제공해야하는가도 Service의 책임이라는 것입니다.

다만 이 고민은 3-4단계 요구사항은 아니라고 판단해 반영하지 않았습니다.(변경 코드가 너무 많아질 것도 같고요..!) 현재 미션에서 JPA에 관련된 핵심 목표를 모두 달성하면 더 고민하여 리팩터링해볼 생각입니다! 따라서 이 코멘트는 간단하게나마 지노의 생각을 듣고 싶어 작성합니다☺️

hyunssooo replied:
> 현재 제 생각은, 어떤 정보를 제공해야하는가도 Service의 책임이라는 것입니다.

저는 조금 다르게 생각합니다. 어떤 정보를 제공하는지는 service를 사용하는 api 혹은 다른 클라이언트의 책임이라고 생각합니다. 대부분 한 api 가 한 service의 메서드를 사용하기 때문에 service의 메서드 리턴 값을 거의 그대로 api 의 응답으로 나가겠지만 엄밀히 따지면 저는 정확한 가공에 대한 책임은 api 에 있다고 생각해요. 그 이유는 api가 view의 추상화라고 생각하기 때문입니다!

yuyoungrhee replied:
완전 이해했습니다. 이에 관련해서 크루들과도 논의해보면서 생각이 정리되었어요.

**Service는 유즈케이스를 완성하는 데 필요한 정보를 반환한다**
여기서 말하는 “모든 정보”는 도메인 개체의 모든 필드가 아니라, 유즈케이스가 요구하는 의미 있는 전체 정보입니다.

**정리**
- 유즈케이스: Service의 결과로 Reservation에 관한 정보(ReservationResult)를 반환한다 -> Reservation에 관한 정보는 모두 포함해서 준다. 그것이 요구사항이니까
- view: Controller에서는 화면에 보여줄 정보를 뽑아서 반환한다. -> ReservationResult에서 memberName, themeName 등을 뽑아서 반환한다.

+ 추가로 고민했던 것
그렇다면 Member의 password도 Service에서 같이 반환해야할까? -> 아니다. 요구사항은 Member에 대한 정보를 반환한다 -> Member에 대한 정보에 password가 포함될 필요는 없다. 민감한 정보 노출을 시키지 않는 것은 Service에서 결정할 수 있음.

YuyoungRhee committed:
fix(GlobalExceptionHandler): NotFoundException에서 400이 아니라 404를 반환하도록 수정

YuyoungRhee committed:
fix(GlobalExceptionHandler): 중복되는 예외 선언 수정

hyunssooo commented on src/main/java/roomescape/service/ReservationCreationService.java:
```diff
@@ -0,0 +1,92 @@
+import roomescape.service.param.CreateReservationParam;
+import roomescape.service.result.ReservationResult;
+
+@Service
+@Transactional(readOnly = true)
+public class ReservationCreationService {
+
```
예약 생성에 대한 service가 만들어졌네요! 다른 service들과 다르게 예약만 복잡하기 때문에 별개의 서비스가 나왔을까요? 어떤 기준으로 행위 별로 서비스가 만들어 질 수 있을까요? 멍구가 생각한 기준은 무엇일까요? 이 부분은 역시 정답이 없으니 편하게 의견 부탁드립니다~

yuyoungrhee replied:
Service 분리 기준에 대해서 많이 채용되는 방식을 찾아봤습니다.

1. 하나로 쓰기
    ReservationService 하나로 사용
2. 읽기용 / 쓰기용
    CommandService / QueryService
3. 행위(기능)별 분리 (UseCase 단위라고도 함)
    CreationService / DeletionService / UpdateService

현재 예약서비스는 **1번 방식 + 너무 복잡한 로직은 기능단위로 떼서 분리** << 방식입니다.
명확한 기준이 아닌 어느정도 복잡해지면 분리했다의 **느낌적인 부분이라 일관적이지 않습니다.** 그렇다고 3번을 채용하기에는 가독성이나 응집도가 떨어지는 것도 아닌데 클래스를 과도하게 분리한다는 생각도 들고요..

현재 제 방식을 사용하려면 명확한 기준이 필요하다고 생각됩니다. 예를 들어,
- 특정 기능이 다른 기능과는 다른 의존성을 O개 이상 필요로 함
- 하나의 기능이 코드 OOO줄 이상
- 특정 기능을 테스트하기 위한 테스트케이스가 O개 이상

일관성이 목표라면 다시 ReservationService로 합치는 것에는 큰 거부감이 없는데, 현재 제 방식도 괜찮은 방식인지 궁금합니다.

hyunssooo replied:
이 코멘트 제가 놓쳤네요 ㅠㅠ 팀내에 컨벤션이라면 저는 좋습니다. 다만 복잡하게 섞어서 사용하는 경우는 드물 것 같아요. 하나로 ReservationService로 사용하든 기능 별로 나누던 할 것 같습니다. 읽기 쓰기로 나누는 것은 별도의 얘기라고 생각해요.

yuyoungrhee replied:
Waiting을 나눈 시점에서 ReservationService가 복잡하지 않아 하나로 다시 합쳤습니다! 나중에 복잡해진다면 기능 별로 나누기도 고려해볼 수 있겠네요. 현재는 하나로 써도 충분히 가독성 있는 것 같아요 😀

[refactor(ReservationService): 예약 생성용 서비스 삭제, ReservationService로 통합](https://github.com/woowacourse/spring-roomescape-waiting/pull/255/commits/4b846ac8835f7ebd1cf024e0c9776d7e81ab9a8a)

hyunssooo commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -32,17 +32,17 @@ public ResponseEntity<ReservationTimeResponse> create(
 
     @GetMapping
-    public ResponseEntity<List<ReservationTimeResponse>> findAll() {
-        List<ReservationTimeResult> reservationTimeResults = reservationService.findAll();
+    public ResponseEntity<List<ReservationTimeResponse>> getAll() {
+        List<ReservationTimeResult> reservationTimeResults = reservationService.getAll();
```
전체적으로 find에서 get으로 변경되었네요. find와 get은 어떤 차이가 있어 변경되었을까요??

yuyoungrhee replied:
일종의 네이밍 컨벤션이라고 생각합니다. 기준이 명확하게 존재하거나 정의되어있는 것은 아니지만 이미 오랜 관습으로 자리잡은 것으로 알고 있습니다.

- find: 찾아본다
    - 없을 수도 있다는 의미 내포
    - 존재하지도, 존재하지 않을 수도 있으니 찾아보겠다는 의미
    - 없으면 null 또는 Optional 반환
- get: 가져온다
    - 반드시 존재한다고 가정하고 단순히 가져온다
    - 보장된 존재를 가져온다는 뜻
    - 없으면 예외 (get에서 기대하는 바가 아니므로)

hyunssooo commented on src/main/java/roomescape/service/ReservationCreationService.java:
```diff
@@ -0,0 +1,92 @@
+    public ReservationCreationService(ReservationRepository reservationRepository, MemberRepository memberRepository,
+                                      ThemeRepository themeRepository,
+                                      ReservationTimeRepository reservationTimeRepository,
+                                      ReservationPolicy reservationPolicy) {
```
사소하지만 이런 부분은 코드를 정리해 주어도 좋습니다. :)

yuyoungrhee replied:
모두 줄바꿈해주도록 수정했습니다!

hyunssooo commented on src/main/java/roomescape/domain/ReservationStatus.java:
```diff
@@ -1,7 +1,17 @@
 package roomescape.domain;
 
 public enum ReservationStatus {
+    RESERVED("예약"),
+    WAITING("대기"),
```
> 현재 코드에서는 분리해야될만큼의 분리된 개념은 아니며, Reservation에서 충분히 처리 가능하다. 그러나 본질적으로는 다른 개념으로 느껴지며 얼마든지 분리될 수 있는 개념이라 생각된다.

네 저도 Reservation에서 처리가 가능하고 상태로 가는 것도 좋다고 생각합니다. 말씀처럼 설계에 따라 다를 것 같아요. 분리하신다면 어떤 방법으로 분리 가능할까요? :)

yuyoungrhee replied:
[해당 코멘트](https://github.com/woowacourse/spring-roomescape-waiting/pull/255#issuecomment-2901404335)에서 정리해봤습니다!

hyunssooo commented on src/main/java/roomescape/domain/ReservationTimePolicy.java:
```diff
@@ -0,0 +1,19 @@
+import java.time.LocalTime;
+import org.springframework.stereotype.Component;
+import roomescape.exception.ReservationTimeException;
+
+@Component
+public class ReservationTimePolicy {
```
👍

hyunssooo commented on src/test/java/roomescape/domain/ReservationTest.java:
```diff
@@ -0,0 +1,80 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import roomescape.TestFixture;
+
+class ReservationTest {
+
```
도메인 모델에 대한 테스트도 작성해 주셨네요. 👍

hyunssooo commented on src/test/java/roomescape/domain/ReservationPolicyTest.java:
```diff
@@ -0,0 +1,82 @@
+    private static final String TOO_SOON_RESERVATION_MESSAGE = "예약 시간까지 10분도 남지 않아 예약이 불가합니다.";
+    private static final String DUPLICATE_RESERVATION_MESSAGE = "테마에 대해 날짜와 시간이 중복된 예약이 존재합니다.";
+
+    @ParameterizedTest
+    @CsvSource({
+            "2025-04-23T12:00, 2025-04-23T11:59",
+            "2025-04-23T12:00, 2025-04-22T12:00"
+    })
+    void 지난_날짜_예약은_불가능하다(String fixedNowStr, String reservationDateTimeStr) {
```
```suggestion
    @ParameterizedTest
    @CsvSource({
            "2025-04-23T12:00, 2025-04-23T11:59",
            "2025-04-23T12:00, 2025-04-22T12:00"
    })
    void 지난_날짜_예약은_불가능하다(LocalDateTime fixedNowStr, LocalDateTime reservationDateTimeStr) {
```

LocalDateTime 으로 받을 수 있을 것 같아요.

yuyoungrhee replied:
그렇네요 훨씬 간결해질 것 같습니다.
반영했습니다!

hyunssooo reviewed (changes requested):
안녕하세요. 멍구! 
이번 step 잘 진행해 주셨습니다! 
몇 가지 가벼운 코멘트 남겼습니다! 확인 부탁드려요!

YuyoungRhee committed:
refactor: Waiting 도메인 따로 분리

YuyoungRhee committed:
refactor: 메서드명 변경 및 코드 정리

YuyoungRhee committed:
refactor: Controller에서 응답을 가공하도록 수정

- 기존: Service에서 필요한 응답 가공
- 변경: Controller에서 Service의 정보를 토대로 필요한 응답 가공

YuyoungRhee committed:
refactor: Reservation과 Waiting의 공통된 부분을 Booking으로 추상화

YuyoungRhee committed:
feat: 몇번째 대기인지 순서를 표시해주도록 구현

YuyoungRhee committed:
refactor: 각 계층의 dto를 dto 패키지로 묶음

YuyoungRhee committed:
refactor: view에 보여줄 status 문자열 결정을 response dto에서 하도록 로직 이동

YuyoungRhee committed:
test: JpaWaitingRepository에 대한 테스트 추가

YuyoungRhee committed:
refactor: equals & hashCode 재정의 삭제

- 비즈니스 로직에서 equals 비교를 쓰지 않으므로 현재 필요 없음

YuyoungRhee committed:
refactor: 사용하지 않는 코드 및 클래스 삭제

YuyoungRhee committed:
feat: request에 대한 검증 추가

- jakarta.validation 의존성 추가

YuyoungRhee committed:
refactor(reservation): 예약 검색 파라미터를 dto로 포장

YuyoungRhee commented:
## Reservation - Waiting 분리 구조

### 코드 구조

Reservation과 Waiting은 Booking을 상속받고 있습니다.

<img width="955" alt="image" src="https://github.com/user-attachments/assets/f831eb15-4cf6-4218-86e7-dca086962266" />

### 테이블 구조

테이블에는 Booking 정보가 각각 들어갑니다.

<img width="281" alt="스크린샷 2025-05-22 오후 7 16 18" src="https://github.com/user-attachments/assets/ade98333-8f84-498a-a9f5-862195af11d1" />

<img width="396" alt="스크린샷 2025-05-22 오후 7 16 29" src="https://github.com/user-attachments/assets/34450360-79aa-462d-8c1b-09eff78ea6c9" />

### 느낀 점

**장점**
- 대기 순번을 id로 계산하는 것이 id 생성 전략에 종속적이라고 생각해 waitingStartedAt이라는 필드를 추가했습니다. 해당 필드는 Waiting에서만 의미가 있으므로 분리 후 적용이 더 편했습니다.
- 이처럼 현실에서는 "예약"과 "대기"를 명확히 구분하여 요구사항이 추가됩니다. Reservation에서 status로 관리하는 경우, 변경 요인이 여러가지가 되면서 응집도가 낮아집니다. 이 부분에서 기존 설계의 단점을 느꼈습니다.
    - "예약"과 "대기"가 다르다고 느낀 이유: ex) 대기 자동 예약, 대기 취소 검증, 대기 순위 계산 등은 예약과는 무관함 (ReservationRepository에서 모두 처리하고 있어서 상당히 메서드가 많아졌음)

**단점**
- 클래스 개수가 많아집니다. 설계적 단점은 아니지만, 코드는 사람이 본다는 점에서 확실한 단점으로 느껴졌습니다.
- 두가지를 동시에 요청하거나, 동시에 필요한 로직이 존재합니다. 애초에 예약과 대기는 개념적으로도 결합이 있기 때문에 코드에서도 결합이 생깁니다.
    - 예를 들어 ReservationService에서 reservationRepositor와 waitingRepository가 필요하며
    (예약 취소시 대기자가 있다면 자동예약)
    - MyReservationService에서도 reservationRepository와 waitingRepository를 둘다 참조합니다
    (나의 예약 조회 시 reservation과 waiting을 모두 조회해야함)

YuyoungRhee committed:
fix: 프론트 코드 대기 생성 경로 수정

YuyoungRhee committed:
refactor: Reservation과 Waiting이 BookingInfo를 @Embeddable로 가지고 있는 구조

YuyoungRhee commented:
지노 리뷰 요청 후 수정 죄송합니다!
Reservation과 Waiting을 분리하는 다른 방법이 생각나 적용해보았습니다.

## 상속 구조가 아닌 조합으로 해결하는 방법

<img width="853" alt="스크린샷 2025-05-23 오후 2 42 42" src="https://github.com/user-attachments/assets/3bc74cb1-1d78-498f-bf65-d441a83727c6" />


- 테이블 구조는 같습니다.
- 조합 방식 (`@Embeddable`)

**장점:**
- 조합의 유연성: "has-a" 관계로 더 자연스러움
- 상속의 단점인 강결합에서 벗어남
- 각각의 요구사항이 달라져 점점 필드가 달라지는 상황에서도 독립적 

**단점:**
- 쿼리 복잡성: r.bookingInfo.member.id 형태 필요 (기존: r.member.id)
- 쿼리 메서드 제한: findByBookingInfoMemberId() 같은 긴 이름
- 위임 메서드: getter들을 일일이 위임해야 함

---
Repository 의 불편함보다 도메인 설계의 명확성이 중요하다고 생각해서 조합으로 변경해봤습니다. 도메인 vs Jpa의 간편한 사용 의 트레이드 오프인 것 같은데, 의견 들려주시면 감사하겠습니다! 각 설계가 모두 이해가 되어서 유독 결정하기 어렵네요.

hyunssooo commented on src/main/java/roomescape/domain/Waiting.java:
```diff
@@ -0,0 +1,74 @@
+import org.springframework.data.annotation.CreatedDate;
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
+@Entity
+@EntityListeners(AuditingEntityListener.class)
+public class Waiting {
+
```
`Waiting`가 별도의 모델로 분리되었네요. 상태를 필드로 가지는 것과 관점의 차이라고 생각합니다. :) 저 또한 두 방식 모두 괜찮다고 생각해요. 
저희가 보통 예약을 하는 경우 `예약이 대기중이다.` 혹은 `예약이 확정 대기중이다.`  같은 말을 많이 사용했던 것 같은데요. 이 맥락에서 저는 `Waiting` 또한 예약의 일부라고 생각해요. 그리고 `Waiting` 은 예약의 상태로 볼 수 있을 것 같아요. 그렇다면 `WaitingReservaion` 이라는 이름도 저는 괜찮을 것 같아요. 클래스로 Reservaion의 상태를 나타낼 수 있을 것 같아서요. 그렇다면 Reservation 이라는 클래스는 조금 어색해 질 것도 같네요! 
제가 짧게 나마 생각한 구조에서는 `Waiting#toReservation` 도 어색한 것 같아요. 다음 상태로 가기 위해서 confirm 같은 메서드가 나와야 하지 않을까 싶어요! 
만약 이 구조라면 repository, service는 어느 단위로 나올 수 있을까요?

hyunssooo replied:
사소할 수 있지만 해당 클래스도 테스트가 작성되면 좋을 것 같아요. :)

yuyoungrhee replied:
> 이 맥락에서 저는 Waiting 또한 예약의 일부라고 생각해요. 그리고 Waiting 은 예약의 상태로 볼 수 있을 것 같아요. 그렇다면 WaitingReservaion 이라는 이름도 저는 괜찮을 것 같아요. 클래스로 Reservaion의 상태를 나타낼 수 있을 것 같아서요. 그렇다면 Reservation 이라는 클래스는 조금 어색해 질 것도 같네요!

- 동의합니다.. 실제 요구사항에서도 `예약 대기 목록 조회`같이 대기를 예약의 상태로 보는 뉘앙스가 강합니다. 여기서 WaitingReservation이라고 이름을 지을 경우 "예약의 상태로 본다"라는 관점을 명확하게 전달하게 됩니다. 다만 저는 이 부분에서 설계 및 네이밍 전체가 헷갈릴 수 있을 것 같아 상당히 고민되네요.
- 현실 요구사항에서는 `예약이 대기중이다`, `예약이 확정 대기중이다` 같은말을 쓰고 있지만, Waiting / Reservation을 분리한 김에 코드로는 조금 다르게 풀어내도 될까요? 간결함과 명확성을 위해서요.
    - Waitng은 예약에 대한 대기가 아니라, 해당 테마-날짜-시간에 대한 "승인 전 상태"를 의미한다고 정해볼 수도 있을 것 같아요. 이런 관점은 어떻게 생각하시나요? 
    - 이렇게되면 테마-날짜-시간을 하나의 슬롯으로 본다는 개념이 코드에도 표현되어야할 것 같아서 BookingSlot으로 Waiting과 Reservation에 표현해줬습니다. 
    [refactor: BookingInfo를 BookingSlot으로 변경하여 예약 슬롯 개념 명확화](https://github.com/woowacourse/spring-roomescape-waiting/pull/255/commits/f59756eb7de39f05b3b4de76b9685adf90869202)

`+` confirm 메서드명은 적용했습니다! 실제 요구사항이 승인 / 거절인만큼 더 와닿는 메서드명인 것 같아요

yuyoungrhee replied:
> 만약 이 구조라면 repository, service는 어느 단위로 나올 수 있을까요?

### Repository: 도메인 중심
**도메인 단위 Repository**
- 한 도메인의 CRUD 책임 담당
    - ex) ReservationRepository, WaitingRepository

**조회 특화 Query Repository**
- 여러 도메인 결합 / 화면 응답에 최적화된 복합 조회
    - ex) BookingQueryRepository

### Service: 기능(Use Case) 중심
**단일 도메인 기능 분리**
- 생성, 수정, 삭제 등 역할 별 기능 중심
    - ex) ReservationCreationService, ReservationCancelService

**사용자/관리자 등 역할에 따른 흐름 구분**
- ex) MyPageService, AdminReservationService

---
우선 현재 적용되어있는 구조 위주로 정리해보면 이와 같습니다. 혹시 어떤 의도로 코멘트 주셨는지 알 수 있을까요? 더 찾아볼 키워드 등이 있을까요?

hyunssooo replied:
말씀하신 도메인 단위라는 것이 entity라면 entity 마다 repository가 생길 것 같아요. 연관관계가 있는 entity들도 모두 repository가 필요한가 싶어서요. 예를 들어 Reservation는 member와 연관관계가 있는데 각각 repository가 존재하고 있어서요. 또한 지금 구조에서는 없지만 oneToMany 관계에서는 Many 쪽의 Repository가 필요할까? 라는 생각이 들어서 남긴 코멘트였습니다. :) 이 부분은 지금은 고민하지 않더라도 추후에 조금씩 고민해 보시면 좋을 것 같아요! 지금은 따로 코드 변경이 없어도 좋습니다!

hyunssooo replied:
> 현실 요구사항에서는 예약이 대기중이다, 예약이 확정 대기중이다 같은말을 쓰고 있지만, Waiting / Reservation을 분리한 김에 코드로는 조금 다르게 풀어내도 될까요? 간결함과 명확성을 위해서요.
> Waitng은 예약에 대한 대기가 아니라, 해당 테마-날짜-시간에 대한 "승인 전 상태"를 의미한다고 정해볼 수도 있을 것 같아요. 이런 관점은 어떻게 생각하시나요?
> 이렇게되면 테마-날짜-시간을 하나의 슬롯으로 본다는 개념이 코드에도 표현되어야할 것 같아서 BookingSlot으로 Waiting과 Reservation에 표현해줬습니다.

간결함과 명확성은 누구의 기준일까요? 개인적으로 코드와 일반적으로 사용하는 용어들은 통일되는 것이 좋다고 생각합니다. 제품을 만드는 과정에서 제품을 만들기 위한 용어(기획)와 제품을 개발하는 용어(개발)의 용어가 별도로 존재하는 것은 커뮤니케이션 비용(번역 비용)을 늘린다고 생각해요. 그런 의미에서 README에 요구 사항을 조금 다듬어 보는 것은 어떨까요? 개발자, 미션을 위한 요구 사항이 아닌 제품 관점에서요! 

조금 우려되는 것은 미션의 목표(jpa) 와 조금 멀어지는 기분이 들긴하네요. 이 부분은 멍구가 선택해 주시면 좋을 것 같습니다!

yuyoungrhee replied:
도메인 용어를 명확히 하기위해 README를 정리해보았습니다.
여기서 Waiting이 "같은 테마,날짜,시간의 Reservation에 대한 대기"가 아니라 "테마, 날짜, 시간에 대한 예약이 아직 승인되지 않은 상태"를 의미하도록 개념을 정리했습니다.

[docs(README) 도메인 개념 정리, 구현 기능 정리 추가](https://github.com/woowacourse/spring-roomescape-waiting/pull/255/commits/8e5ceb0985fc65b56f7fed661fdba697e5c22912)

yuyoungrhee replied:
> 또한 지금 구조에서는 없지만 oneToMany 관계에서는 Many 쪽의 Repository가 필요할까? 

지노 말대로 지금 구조는 아니지만 만약 Member가 `List<Reservation>`을 가지게 된다면 Reservation을 처리하는 주체가 Member가 될 것 같아요. member.add(reservation) 등과 같이요. 만약 모든 reservation의 기능(CRUD 등)이 Member를 거쳐야한다면  ReservationRepository는 필요 없겠네요.

따라서 Repository는 도메인 별로 모두 존재한다의 정의는 옳지 않겠네요,
**Repository는 도메인의 독립적 관리 책임이 있을 때 필요하다** 정도면 괜찮을까요?
- 단독으로 저장/조회/수정/삭제돼야 할 도메인(예약, 회원, 테마 등)은 각각 Repository가 필요하다.

hyunssooo commented on src/main/java/roomescape/domain/Waiting.java:
```diff
@@ -0,0 +1,74 @@
+        return new Waiting(null, member, date, time, theme);
+    }
+
+    public Reservation toReservation() {
+        return Reservation.create(getMember(), getDate(), getTime(), getTheme());
+    }
```
BookingInfo를 그대로 사용하지 않은 이유가 있을까요?

yuyoungrhee replied:
처음에는 BookingInfo가 Waiting과 Reservation 안에서 캡슐화되어 관리되는 VO라고 생각하여 외부에서 BookingInfo를 만들어 넘겨주는 것이 아니라, Reservation이 스스로 정보를 가지고 `new BookinInfo(넘겨받은 정보)`를 하도록 하는 구조였습니다.

코멘트를 통해 다시 생각해봤는데, 어차피 도메인에서 공유되는 VO고 재사용성과 의미적으로 전달력을 생각해봤을 때 bookingInfo를 그대로 넘겨 만드는 것도 좋은 방법이라 생각해 수정했습니다. 또한 BookingInfo에 대한 생성 책임을 여러군데가 아닌 오직 BookingInfo에서 담당하면서 책임분리면에서도 유리하다고 생각합니다.
[refactor: Reservation과 Waiting이 BookingInfo를 인자로 받도록 생성자 리팩터링](https://github.com/woowacourse/spring-roomescape-waiting/pull/255/commits/0e3be183a674ab2385fed8c86473f35dd160828c)

❗️현재는 BookingSlot으로 설계가 바뀌었습니다..!

hyunssooo commented on src/main/java/roomescape/domain/repository/WaitingRepository.java:
```diff
@@ -0,0 +1,24 @@
+
+    List<WaitingWithRankData> findWaitingsWithRankByMemberId(Long memberId);
+
+    List<Waiting> findAll();
+}
```
상태와 상관없이 예약을 보여줘야 한다면 어떤 방법이 있을까요?

yuyoungrhee replied:
대기와 예약을 함께 조회하는 책임을 담당하는 BookingQueryRepository를 만들어 사용할 수 있을 것 같아요. 마침 이번 리팩터링을 진행하면서 만들어보았습니다.

기존에는 다음과 같이 Service에서 각각 조회 후 조합하고 있었는데요,

```java
    public List<MemberBookingResult> getMyBookings(Long memberId) {
        List<Reservation> reservations = reservationRepository.findByMemberId(memberId);
        List<WaitingWithRankData> waitings = waitingRepository.findWaitingsWithRankByMemberId(
                memberId);

        return Stream.concat(
                        reservations.stream().map(MemberBookingResult::from),
                        waitings.stream().map(data -> MemberBookingResult.from(data.waiting(), data.rank() + 1))
                ).sorted(Comparator.comparing(MemberBookingResult::reservationDateTime))
                .toList();
    }

```

만약 페이징을 하게 된다면 불가능한 방법으로, 예약과 대기를 한꺼번에 DB에서 조회해오는 로직의 필요성을 느꼈습니다. 이를 BookingQueryRepository라고 두고 구현해보았습니다.
[refactor: 내 에약과 대기를 모두 조회하는 로직을 DB에서 한꺼번에 가져오도록 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/255/commits/075de9963cab62aed157b96b350d0a6e558c27b7)

한 가지 고민은 nativeQuery를 사용하여 복잡한 로직을 처리 중인데, 한 번에 쿼리를 이해하기에는 어려움이 있을 것 같습니다. 가독성을 좋게하는 방법이 있을까요? 아니면 nativeQuery 이용시 현재 코드가 최선일까요?

hyunssooo replied:
점점 조회가 복잡해 지는 것 같은데요. 조회는 비즈니스 로직과 별도로 생각하셔도 괜찮을 것 같아요. 복잡한 쿼리가 발생하는 이유는 요구 사항에 보여질 데이터가 복잡하기 때문이라고 생각해요. BookingQueryRepository 도 좋습니다. 이 경우에는 조회 모델이 별도로 나오겠네요!

hyunssooo commented on src/main/java/roomescape/service/WaitingService.java:
```diff
@@ -0,0 +1,94 @@
+
+        return new WaitingComponents(member, theme, reservationTime);
+    }
+
+    private record WaitingComponents(Member member, Theme theme, ReservationTime time) {
+    }
+
```
별도로 클래스를 분리하지 않은 이유가 있을까요?

yuyoungrhee replied:
클래스를 오가는 record가 아니라 WaitingService 클래스 내부에서만 사용되는 **내부 메서드 사이에서** 데이터 묶음으로 전달해주기 위해 편의상 생성한 클래스로, 외부에 공개할 이유가 없다고 생각했습니다. 단순한 파라미터 묶음용이며, 일회성인 record입니다.

hyunssooo reviewed (changes requested):
안녕하세요. 멍구! 
분리해 주신 내용 잘 봤습니다! 
가볍게 제 생각 남겼는데요! 확인 부탁드려요!

YuyoungRhee committed:
feat(waiting): 예약 대기 시작 시간이 같아도 id로 추가 정렬을 하도록 구현

YuyoungRhee committed:
refactor: 내 에약과 대기를 모두 조회하는 로직을 DB에서 한꺼번에 가져오도록 수정

- 기존: Service에서 각각 조회 후 조합
- 변경: JpaBookingQueryRepository에서부터 한번에 가져옴

YuyoungRhee committed:
refactor: Reservation과 Waiting이 BookingInfo를 인자로 받도록 생성자 리팩터링

- BookingInfo를 외부에서 생성해 주입하도록 변경하여 재사용성과 테스트 용이성 향상
- 생성자 또는 팩토리 메서드에서 BookingInfo를 바로 받도록 구조 변경

YuyoungRhee committed:
refactor: 테스트 추가 및 메서드명 변경

YuyoungRhee committed:
refactor: BookingInfo를 BookingSlot으로 변경하여 예약 슬롯 개념 명확화

- Waiting과 Reservation에서 공통으로 사용하던 BookingInfo의 이름을 BookingSlot으로 변경
- member 정보를 Booking에서 분리하여, BookingSlot은 테마/날짜/시간만을 표현하도록 재구성

hyunssooo reviewed (changes requested):
안녕하세요. 멍구 
몇 가지 가벼운 제 생각 남겼습니다. 
확인 부탁드려요!

YuyoungRhee committed:
refactor(ReservationService): 예약 생성용 서비스 삭제, ReservationService로 통합

YuyoungRhee committed:
refactor: 모든 예약 조회 API 및 예약 필터링 API를 관리자용 Controller로 이동

YuyoungRhee committed:
docs: RestDocs를 활용한 API 명세 작성

YuyoungRhee committed:
docs(README) 도메인 개념 정리, 구현 기능 정리 추가

YuyoungRhee commented:
지노 리뷰 감사합니다! 설계 정리에 도움 많이 되었어요. 제가 어떤 식으로 Waiting을 설계하고자 했는지 혼자만 이해하는 것이 아니라, 동료 개발자나 기획자, 페어 등과 같이 제품을 만들고 사용하는 모든 사람들이 이해할 수 있도록 공유해야한다는 중요성을 깨달았습니다.

소소하게 다음과 같은 리팩터링 진행했습니다.
- admin만 가능한 예약 목록 조회 API를 `AdminReservationController`로 이동
- 테스트 정리 및 구체화(name 등 의미없는 단어가 아닌 알아보기 쉬운 단어 사용("멍구" 등))
- README 정리
- API 문서화

hyunssooo commented on README.md:
```diff
@@ -1,307 +1,79 @@
-**id**: 예약 식별자
+- 기능
+  - 관리자가 시간을 추가/삭제/조회하는 API
+  - 사용자가 테마,날짜에 따른 예약 가능한 시간을 조회하는 API
 
-**name**: 예약자명
+### Reservation(예약됨)
 
-**date**: 예약 날짜
+- 기능
+  - 관리자가 예약을 조회(+필터링)/추가/삭제하는 API
+  - 사용자가 BookingSlot(테마-날짜-시간)을 선택하여 예약을 등록하는 API
```
api 라는 단어는 다른 직군이 이해할 수 있을까요? 최대한 다른 직군도 함께 이해할 수 있는 단어를 사용해 보아도 좋을 것 같아요.

hyunssooo reviewed (approved):
안녕하세요. 멍구! 
이번 미션 진행하신다고 정말 고생 많으셨습니다! 
가벼운 코멘트 하나 남겼어요! 추후에 고민해 보셔도 좋을 것 같아요! 
함께 고민이 필요한 부분이 있으시면 언제든 DM 남겨주셔도 좋습니다!
